contracts/serverless-v2/Gateways.sol

```diff
@@ -0,0 +1,685 @@
+        for (uint256 index = 0; index < _requestChains.length; index++) {
+            RequestChain memory reqChain = _requestChains[index];
+            uint256 chainId = _chainIds[index];
+            requestChains[chainId] = reqChain;
+
+            emit ChainAddedGlobal(chainId, reqChain.contractAddress, reqChain.httpRpcUrl, reqChain.wsRpcUrl);
+        }
+    }
+
+    function _removeChainGlobal(
+        uint256[] memory _chainIds
+    ) internal {
+        if(_chainIds.length == 0)
+            revert GatewaysInvalidLength();
+        for (uint256 index = 0; index < _chainIds.length; index++) {
+            uint256 chainId = _chainIds[index];
+            delete requestChains[chainId];
+
+            emit ChainRemovedGlobal(chainId);
+        }
+    }
+
+    function setPaymentPool(address _paymentPoolAddress) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        paymentPool = _paymentPoolAddress;
+    }
+
+    function whitelistEnclaveImage(
+        bytes memory PCR0,
+        bytes memory PCR1,
+        bytes memory PCR2
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) returns (bytes32, bool) {
+        return _whitelistEnclaveImage(EnclaveImage(PCR0, PCR1, PCR2));
+    }
+
+    function revokeEnclaveImage(bytes32 imageId) external onlyRole(DEFAULT_ADMIN_ROLE) returns (bool) {
+        return _revokeEnclaveImage(imageId);
+    }
+
+    function addChainGlobal(
+        uint256[] memory _chainIds,
+        RequestChain[] memory _requestChains
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _addChainGlobal(_chainIds, _requestChains);
+    }
+
+    function removeChainGlobal(
+        uint256[] memory _chainIds
+    ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+        _removeChainGlobal(_chainIds);
+    }
+
+    //-------------------------------- Admin methods end ----------------------------------//
+
+    //-------------------------------- internal functions start ----------------------------------//
+
+    function _registerGateway(
+        bytes memory _attestationSignature,
+        IAttestationVerifier.Attestation memory _attestation,
+        uint256[] memory _chainIds,
+        bytes memory _signature,
+        uint256 _stakeAmount,
+        uint256 _signTimestampInMs,
+        address _owner
+    ) internal {
+        // attestation verification
+        _verifyEnclaveKey(_attestationSignature, _attestation);
+
+        address enclaveAddress = _pubKeyToAddress(_attestation.enclavePubKey);
+        // signature check
+        _verifyRegisterSign(_owner, _chainIds, _signTimestampInMs, _signature, enclaveAddress);
+
+        if(gateways[enclaveAddress].owner != address(0))
+            revert GatewaysGatewayAlreadyExists();
+
+        for (uint256 index = 0; index < _chainIds.length; index++) {
+            if(requestChains[_chainIds[index]].contractAddress == address(0))
+                revert GatewaysUnsupportedChain();
+        }
+
+        // TODO: check missing for validating chainIds array for multiple same chainIds
```

Comment:
Actually, unique chain IDs in a list can be ensured by the enclave. Because, it signs the list. Contract can always be sure that list contain unique chain IDs.